### ISSUE TITLE
Raising GEOS "notice" messages as warnings not working?

### ISSUE THREAD

**Issue by jorisvandenbossche** [shapely/shapely#1283]
Moved from https://github.com/pygeos/pygeos/issues/214, and related to the logging discussion at #998

Currently, we have the intention to use warnings for the "notice" messages of GEOS (and exceptions for "error" messages):

https://github.com/shapely/shapely/blob/48cf81e5ba3e48b779e81be543ab42858385c37d/src/geos.h#L130

https://github.com/shapely/shapely/blob/48cf81e5ba3e48b779e81be543ab42858385c37d/src/geos.h#L64-L67

Looking into the GEOS source code to see what these "notices" are used for, I *think* it is actually only used for the `is_valid` operation (see https://github.com/libgeos/geos/blob/27516e0bdb7526699c00fef328c61561ab7a4b47/capi/geos_ts_c.cpp#L688). All other errors are sent to the "error" handler. 

From my reading of the GEOS code, I expected that this would raise a warning:

```python
import warnings
warnings.simplefilter("always")

import shapely
import shapely.geometry
coords = [(0, 0), (0, 2), (1, 1), (2, 2), (2, 0), (1, 1), (0, 0)]
bowtie = shapely.geometry.Polygon(coords)
assert bowtie.is_valid is False
assert shapely.is_valid_reason(bowtie) == "Ring Self-intersection[1 1]"
```

However, running the above code does not actually raise a warning on the main branch.

As comparison, the equivalent with Shapely 1.8 and setting up the logger to print to stdout, _does_ show the message from GEOS:

```python
import sys
import logging

logger = logging.getLogger('shapely.geos')
logging_handler_out = logging.StreamHandler(sys.stdout)
logger.addHandler(logging_handler_out)
logger.setLevel(logging.INFO)

from shapely.geometry import Polygon

coords = [(0, 0), (0, 2), (1, 1), (2, 2), (2, 0), (1, 1), (0, 0)]
bowtie = Polygon(coords)

In [9]: bowtie.is_valid
Ring Self-intersection at or near point 1 1
Out[9]: False
```

**Comment by jorisvandenbossche** [comment]
There are two issues here:

- Setting the warning with `PyErr_WarnEx` in `GEOS_FINISH` doesn't seem to actually work. Setting an error with `PyErr_SetString` does work though. 
  I debugged the snippet above, and I can confirm that it actually gets to calling `PyErr_WarnEx` with the correct warning message (so the notice handler itself registered with GEOS is working). I am not familiar enough with the Python C API or numpy C API to know if there is something more needed to get Python to show the warning. 
- Do we actually want the warning? 
  As argued in https://github.com/pygeos/pygeos/issues/214#issuecomment-708985216, having `is_valid` warn about invalid geometries is not that useful (you are using it exactly to check whether there are invalid geometries) and we probably don't want a flood of warnings. So if we keep the notice->warning handling, we should see if we can consolidate the warnings (but first need to fix the warning handling itself, of course)

**Comment by caspervdw** [comment]
My preferred option would be disabling the notice handler altogether. I don’t see a use case in the is_valid operation, because we also have is_valid_reason if users are interested in the message.

**Comment by jorisvandenbossche** [comment]
I agree that a warning would not be useful in this specific case of `is_valid`, and so as long that's the only place where GEOS uses a notice, it indeed doesn't seem worth the time to try to fix the notice handling.
